### ISSUE TITLE
Does `stop()` actually allow jobs to complete?

### ISSUE THREAD

**Issue by shivak** [graphile/worker#321]
As far as I can tell, the desired behavior of `stop()` is to disallow new jobs, wait for currently-running jobs to complete, and to release all pools/resources. But it seems like `stop()` actually interrupts jobs. In the following example, whether `noHandleSignals` is set to true or false, the job never completes normally.  

Is there a small working example of jobs managing their own completion upon SIGINT/SIGTERM?
 
```javascript
import sleep from 'sleep-promise';
import {quickAddJob, run} from 'graphile-worker';

// simple task which logs "done" when it completes normally
async function task(payload, helpers) {
  const {logger} = helpers;
  let terminate = false;
  process.once('SIGINT', async (signal) => {
    logger.warn("terminating job");
    terminate = true;
  });

  while (!terminate) { 
    await sleep(1000); 
    logger.info("processing...");
  }

  logger.info("done");
}
await quickAddJob({}, 'task');

// custom handler should allow task to complete?
const noHandleSignals = true;
let runner = await run({
  noHandleSignals,
  taskList: {task: task}
});
if (noHandleSignals) { 
  process.once('SIGINT', async () => { 
    console.log("terminating runner");
    await runner.stop();
  });
}

await runner.promise;
process.exit();
```

**Comment by benjie** [comment]
I ran this script:

```js
import {run, quickAddJob} from 'graphile-worker';
const sleep = (ms) => new Promise(resolve => setTimeout(resolve, ms));

// simple task which logs "done" when it completes normally
async function task(payload, helpers) {
  const {logger} = helpers;
  let terminate = false;
  process.once('SIGINT', (signal) => {
    logger.warn("terminating job");
    terminate = true;
  });

  while (!terminate) { 
    await sleep(1000); 
    logger.info("processing...");
  }

  logger.info("done");
}
await quickAddJob({}, 'task');

// custom handler should allow task to complete?
const noHandleSignals = true;
let runner = await run({
  noHandleSignals,
  taskList: {task: task}
});
if (noHandleSignals) { 
  process.once('SIGINT', async () => { 
    console.log("terminating runner");
    await runner.stop();
  });
}

await runner.promise;
process.exit();
```

Using graphile worker 0.14.0-alpha.0

and the output I got was:

```
$ DATABASE_URL=worker-321 node repro.mjs
[core] INFO: Worker connected and looking for jobs... (task names: 'task')
[job(worker-3852234cd70ab73c1d: task{1})] INFO: processing...
[job(worker-3852234cd70ab73c1d: task{1})] INFO: processing...
[job(worker-3852234cd70ab73c1d: task{1})] INFO: processing...
^Cterminating runner
[job(worker-3852234cd70ab73c1d: task{1})] WARNING: terminating job
[job(worker-3852234cd70ab73c1d: task{1})] INFO: processing...
[job(worker-3852234cd70ab73c1d: task{1})] INFO: done
[worker(worker-3852234cd70ab73c1d)] INFO: Completed task 1 (task) with success (4003.56ms)
```

Seems to be working as designed?

**Comment by shivak** [comment]
Hi Benjie,

Thanks for taking a look. On version 0.13.0, upon either node 18.9.1 or 19.1.0, the following incorrect behavior is observed:

```
$ node halt.mjs
[core] INFO: Worker connected and looking for jobs... (task names: 'task')
[job(worker-5b8f17e0adf5953c5b: task{10684})] INFO: processing...
[job(worker-5b8f17e0adf5953c5b: task{10684})] INFO: processing...
[job(worker-5b8f17e0adf5953c5b: task{10684})] INFO: processing...
[job(worker-5b8f17e0adf5953c5b: task{10684})] INFO: processing...
[job(worker-5b8f17e0adf5953c5b: task{10684})] INFO: processing...
^Cterminating runner
[job(worker-5b8f17e0adf5953c5b: task{10684})] WARNING: terminating job
```

**Comment by benjie** [comment]
Ah, I didn't realise the behavior differed there but yet I can reproduce with 0.13.0. No need to port a fix to 0.13.0, just use 0.14.0-alpha.0 - it's essentially 0.14.0 I just haven't released it yet.

**Comment by shivak** [comment]
Thanks, will do.
